File: clients/kotlin/src/main/kotlin/model/Item.kt
package model

import util.StreamUtil

abstract class Item {
    @Throws(java.io.IOException::class)
    abstract fun writeTo(stream: java.io.OutputStream)
    companion object {
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): Item {
            when (StreamUtil.readInt(stream)) {
                HealthPack.TAG -> return HealthPack.readFrom(stream)
                Weapon.TAG -> return Weapon.readFrom(stream)
                Mine.TAG -> return Mine.readFrom(stream)
                else -> throw java.io.IOException("Unexpected discriminant value")
            }
        }
    }

    class HealthPack : Item {
        var health: Int = 0
        constructor() {}
        constructor(health: Int) {
            this.health = health
        }
        companion object {
            val TAG = 0
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): HealthPack {
                val result = HealthPack()
                result.health = StreamUtil.readInt(stream)
                return result
            }
        }
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            StreamUtil.writeInt(stream, health)
        }
    }

    class Weapon : Item {
        lateinit var weaponType: model.WeaponType
        constructor() {}
        constructor(weaponType: model.WeaponType) {
            this.weaponType = weaponType
        }
        companion object {
            val TAG = 1
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): Weapon {
                val result = Weapon()
                when (StreamUtil.readInt(stream)) {
                0 ->result.weaponType = model.WeaponType.PISTOL
                1 ->result.weaponType = model.WeaponType.ASSAULT_RIFLE
                2 ->result.weaponType = model.WeaponType.ROCKET_LAUNCHER
                else -> throw java.io.IOException("Unexpected discriminant value")
                }
                return result
            }
        }
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            StreamUtil.writeInt(stream, weaponType.discriminant)
        }
    }

    class Mine : Item {
        constructor() {}
        companion object {
            val TAG = 2
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): Mine {
                val result = Mine()
                return result
            }
        }
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
        }
    }
}

File: clients/scala/src/main/scala/model/Item.scala
package model

import util.StreamUtil

sealed trait Item {
    def writeTo(stream: java.io.OutputStream)
}
object Item {
    case class HealthPack(health: Int) extends Item {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, HealthPack.TAG)
            StreamUtil.writeInt(stream, health)
        }
    }
    object HealthPack {
        val TAG: Int = 0
        def readFrom(stream: java.io.InputStream): HealthPack = HealthPack(
            StreamUtil.readInt(stream)
            )
    }

    case class Weapon(weaponType: model.WeaponType) extends Item {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, Weapon.TAG)
            weaponType.writeTo(stream)
        }
    }
    object Weapon {
        val TAG: Int = 1
        def readFrom(stream: java.io.InputStream): Weapon = Weapon(
            model.WeaponType.readFrom(stream)
            )
    }

    case class Mine() extends Item {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, Mine.TAG)
        }
    }
    object Mine {
        val TAG: Int = 2
        def readFrom(stream: java.io.InputStream): Mine = Mine(
            )
    }

    def readFrom(stream: java.io.InputStream): Item = {
        StreamUtil.readInt(stream) match {
            case HealthPack.TAG => HealthPack.readFrom(stream)
            case Weapon.TAG => Weapon.readFrom(stream)
            case Mine.TAG => Mine.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected discriminant value")
        }
    }
}
